### DISCUSSION TITLE
Kubernetes setup

### DISCUSSION THREAD

**Discussion by chevdor** [inventree/InvenTree#2658] [General]
Hello,

I am using a K8s cluster and I am trying to deploy Inventree. I am hitting a few issues that are non K8s related.
I am using `inventree/inventree:stable`.

The main issue seems to be that the DB is not created automatically.
I am using postgres, postgres is running, the db is created (I did it manually), the user is created (manually) and the access is correct, nonetheless, the inventree db remains empty (= no tables).

I ended up:
- creating the db manually
- creating the user manually
- granting the user access

For now I am "just" trying to get it to run but I am a little surprised of the extra effort related to setting up the DB. Usually, for similar projects, as long as the db credentials are ok, the server container checks if the db exsists, and create it if not.

Please correct me if I am wrong but I would expect the following to be done automatically provided the credentials passed are correct:
- db gets created
- provided users gets granted access to the db
- tables & Co are setup and initialized
- migrations are run

Since I don't see any table, I guess something went wrong and I would like to understand wheter I should initially run the migrations manually, or if this is supposed to be handled and I should look elsewhere.

I did turn on debugging but I am not seeing much more.

In a nutshell, the server tells me it spwans workers and they timeout.
The worker start, confirm the db connection, then fail as shown below:

<details>
<summary>Worker logs</summary>

```
Waiting for database...
Database connection sucessful!
Traceback (most recent call last):
  File "/home/inventree/.local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.UndefinedTable: relation "label_stockitemlabel" does not exist
LINE 1: SELECT (1) AS "a" FROM "label_stockitemlabel" WHERE "label_s...
                               ^
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "manage.py", line 22, in <module>
    execute_from_command_line(sys.argv)
  File "/home/inventree/.local/lib/python3.8/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/home/inventree/.local/lib/python3.8/site-packages/django/core/management/__init__.py", line 395, in execute
    django.setup()
  File "/home/inventree/.local/lib/python3.8/site-packages/django/__init__.py", line 24, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/home/inventree/.local/lib/python3.8/site-packages/django/apps/registry.py", line 122, in populate
    app_config.ready()
  File "/home/inventree/InvenTree/label/apps.py", line 38, in ready
    self.create_stock_item_labels()
  File "/home/inventree/InvenTree/label/apps.py", line 114, in create_stock_item_labels
    if StockItemLabel.objects.filter(label=filename).exists():
  File "/home/inventree/.local/lib/python3.8/site-packages/django/db/models/query.py", line 808, in exists
    return self.query.has_results(using=self.db)
  File "/home/inventree/.local/lib/python3.8/site-packages/django/db/models/sql/query.py", line 552, in has_results
    return compiler.has_results()
  File "/home/inventree/.local/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1145, in has_results
    return bool(self.execute_sql(SINGLE))
  File "/home/inventree/.local/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1175, in execute_sql
    cursor.execute(sql, params)
  File "/home/inventree/.local/lib/python3.8/site-packages/django/db/backends/utils.py", line 98, in execute
    return super().execute(sql, params)
  File "/home/inventree/.local/lib/python3.8/site-packages/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/home/inventree/.local/lib/python3.8/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/home/inventree/.local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/home/inventree/.local/lib/python3.8/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/home/inventree/.local/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: relation "label_stockitemlabel" does not exist
LINE 1: SELECT (1) AS "a" FROM "label_stockitemlabel" WHERE "label_s...
```
</details

The issue is legit since, connecting to the postgres server, I can indeed see that the `inventree` db does not have a single table.

**Comment by chevdor** [comment]
OK, I found some hints and answers in the doc, it suggests the need to connect to the server first and run:
```
invoke update
invoke superuser
```

In the context of k8s, that could be a initContainer. It would be great to have a script in the container that does just that, but only if the db was not found. That would allow running this from the initContainer, no matter what, but the script would only run the invoke on environment that are detected as "new".

Happy to turn that into an issue if you can confirm you agree with the following:

- new script `./setup.sh`
- it connects to postgress using the provided user + creds
- it checks whether the db exists
- if yes, it exits
- if no, it:
  - create db
  - grant user access
  - run the `invoke update` and `invoke superuser`

**Reply by matmair** [reply]
Hi there @chevdor!
At the moment Kubernetes is not a supported deployment method. There are users how did it but there is no infra from us for it.
We would gladly welcome FRs or PRs that make this possible - maybe with sample charts or a gh-repo with a good starting template.

In general ppl. seem to not read the whole docker docs so please make sure to read them before trying anything. We have a single command that needs to run on install/updates (Inv update - this one is safe to run anytime) and one that is needed for creation of a user (Inv ~~superset~~ superuser).

Also we are database agnostic - you can use sqlite3, postgres, mysql, mssql or even Redis, mongo or cosmoDB if you want. So solutions should be also agnostic.

**Reply by chevdor** [reply]
I got a basic version running now but I am still missing a few things.
I am (happily) surprised to see that redis could be used directly. I suppose this provides much better performance. I was actually wondering if redis could be used for caching on top of a db such as postgres.

**Reply by matmair** [reply]
> I am (happily) surprised to see that redis could be used directly. I suppose this provides much better performance. I was actually wondering if redis could be used for caching on top of a db such as postgres.

A few ppl (myself included) deploy with redis. You can very easily use all caching and database things supported by django. Just make a custom settings file, import everything from the original settings and change what file gunicorn and the cluster start with.

**Reply by jesseyu1984918** [reply]
Just curious if things turn out better in 2024? :P

**Reply by chevdor** [reply]
Not on my end, since:
> Kubernetes is not a supported deployment method

I switch to another solution that is more K8s friendly.

**Reply by jesseyu1984918** [reply]
May I know what you have switched to?

**Reply by chevdor** [reply]
I am using Homebox. In the meantime, the project slew down and ended up archived but it has been forked and the new team shows quite some motiviation and activity.

**Reply by wolflu05** [reply]
#6524 might be interesting for k8s and there are also some envs to setup an initial admin user since a while.

**Reply by aep** [reply]
homebox is a bit simplistic unfortunately for our use case (server components).
inventree keeps falling apart unfortunately, its also clearly not intended for production use.
I wonder if such a software is even needed anymore with all the smart sheets like airtable etc

**Reply by matmair** [reply]
InvenTree runs in a lot of production setups. Kubernetes is just not a supported deployment method as we do not have anyone that has experience with it in the team or community.

**Comment by chevdor** [comment]
See also https://github.com/inventree/InvenTree/issues/2659
